### ISSUE TITLE
Model support: meta-llama/Llama-3.2-3B-Instruct

### ISSUE THREAD

**Issue by ftynse** [martin-luecke/rocm-systems#75]
- [x] unmodified model runs on gfx942 OR gfx950
- [ ] relevant kernels extracted (link to the hotswap-testing repository)
- [x] transpilation succeeds on gfx942 OR gfx950
- [x] individual kernels run with synthetic data
- [x] transpiled kernels run and produces result within FP noise boundary on synthetic data (no NaNs)
- [x] whole model runs with transpiled kernels
- [x] relevant throughput metric (token/s, frame/s) measured and reported below (alongside git hashes used)
- [ ] all of the above on BOTH gfx942 AND gfx950

**Comment by tgymnich** [comment]
waiting for model access

**Comment by YashDeshpande25** [comment]
## Llama-3.2-3B-Instruct — gfx950 throughput (warm run2)
GPU: AMD Instinct MI350X (`gfx950`)  

### HIP graphs disabled (`disable_cuda_graph=True`)

| Branch | Generation tok/s | HotSwap / local |
|--------|-----------------:|----------------:|
| Local | 1.143 | — |
| HotSwap | 0.831 | 0.727 |

### Numerical equivalence (local vs HotSwap, warm `run2`)
- **Verdict:** PASS (`numerically_close`)  
- **Token divergence:** 0 / 1 cases (4/4 matching prefix)  
- **`nan_logprobs`:** 0 on all cases (no NaN/Inf in logprob streams)  
- **Max chosen-token logprob diff:** 0.040  
- **Max forward KL:** 6.614e-04  

### Git hashes

| Component | Commit |
|-----------|--------|
| rocm-hotswap-testing (`main`) | `9ad16c2f87c1d3af7f568b986ce7983f9cd1341e ` |
| llvm-project (hotswap / COMGR) | `56beab43b6dabc9d21370a92a2fb799f262e5c80` |
| rocm-systems (HotSwap ROCR adapter) | `7bde349cf160290e278cef0ebad42f8a9c910222` |
| sglang | https://github.com/sgl-project/sglang  (`3d95ca7546fb03227b7c66bd72cca965655899f0`) |

**Comment by ishkool** [comment]
I'll pick this up.

**Comment by ishkool** [comment]
Awaiting access grant for the model.

<img width="1777" height="352" alt="Image" src="https://github.com/user-attachments/assets/a594ee22-c6df-4b3a-a9ba-f755908bb3db" />

**Comment by ishkool** [comment]
## gfx950 validation report — `meta-llama/Llama-3.2-3B-Instruct`

**TL;DR:** The model runs end-to-end through the PyTorch HotSwap runner on gfx950. The native/local branch passes and the **whole-model HotSwap run passes**: **38/38 Triton kernels transpiled** gfx1250→gfx950 (0 raise failures) and greedy generation is **numerically equivalent** to native (0 token-divergent across all prompts; per-step logprob/KL/TVD diffs are exactly `0.0`). Enabling this causal-LM required three targeted workload changes (disable Inductor/Triton autotuning + the static CUDA launcher, pin a static KV cache under `torch.compile`, and rewrite RoPE to avoid a degenerate GEMM). The `phi4`-style library-kernel override path was evaluated and found unusable here (see below).

### Environment

| Component | Value |
|---|---|
| GPU | AMD Instinct **MI350X**, `gfx950` (wave64) |
| Container | `rocm/pytorch:latest`, ROCm **7.2** (`hip 7.2.53211`) |
| `llvm-project` (COMGR) | branch `hotswap` @ `c66543a689994f94e7a8c7043c8c18653cc0b193` |
| `rocm-systems` (ROCR) | `users/mluecke/hotswap-compatibility` @ `26b07419c2e5985aa9874b4448190ccb6cc6286c` |
| `rocm-hotswap-testing` | `users/adespirl/non-embedding-model-groups` @ `cfc62d47436c26a4c18f8d6f361f25dff46799ec` |
| torch / triton | `2.12.0+rocm7.2` / `triton-rocm 3.6.0` |
| transformers | causal-LM (`AutoModelForCausalLM`), `sdpa` attention |
| Source target / target | `gfx1250:32` (wave32) → `gfx950` (wave64) |
| Precision / workload | `bfloat16`, 3 chat prompts, greedy decode, `max_new_tokens 16`, `top_logprobs 5` |

Llama-3.2-3B-Instruct is gated; weights were downloaded locally to `/home/ishkool2/models/Llama-3.2-3B-Instruct` and passed via `--model-path`.

### Runner enablement

Llama fits the existing generic `hf_causal_lm` workload — adding it is a JSON config, not a new Python module. Config: `data/pytorch/models/llama_3_2_3b_instruct.json` (`workload: hf_causal_lm`, `source_target: gfx1250:32`, `target_gfx: gfx950`, `PYTORCH_COMPILE: 1`, equivalence `atol_logprob 5e-3` / `kl_tolerance 1e-3`) plus `llama_3_2_3b_instruct.prompts.json`.

A causal LM run in eager mode dispatches entirely to ATen + rocBLAS/hipBLASLt library kernels and emits **no** JIT Triton kernels, so there is nothing for the transpiler to translate (an eager run passed equivalence but recorded `0/0` HotSwap proof). To exercise the HotSwap path the model is run under `torch.compile` (TorchInductor → Triton), which routes fused pointwise/reduction ops (RMSNorm, RoPE apply, SiLU, softmax, residual adds) through Triton; those kernels are force-compiled for `gfx1250:32` and translated to `gfx950` on load. Matmuls (Q/K/V/O and MLP projections, `lm_head`) still dispatch to the native gfx950 BLAS backend — the same coverage model as the accepted DiT/FLUX runs.

Three workload changes in `runtime/pytorch/workloads/hf_causal_lm.py` were required to make that path actually work under HotSwap:

1. **Static KV cache under compile** — HF `generate()` graph-breaks on its default dynamic cache, so Inductor never compiled the decode graph (first attempt produced an empty Triton cache and `0/0` proof). Passing `cache_implementation="static"` gives a stable graph so Inductor emits Triton.
2. **Disable all Inductor/Triton autotuning** — default lowering attaches multiple candidate configs per kernel and benchmarks them by launching each; a single candidate that fails to translate aborts the whole sweep. Disabled `max_autotune*`, `coordinate_descent_tuning`, `triton.autotune_pointwise`, `triton.autotune_cublasLt`, and `triton.autotune_at_compile_time` so each kernel emits one deterministic config (kernel count 110 → 48).
3. **Disable the static CUDA launcher** (`use_static_cuda_launcher=False`) — it loads kernels through a C extension that bypasses `hipModuleLoadDataEx`, the exact API the HotSwap intercept hooks, so freshly compiled gfx1250 kernels loaded raw on gfx950 → `hipErrorNoBinaryForGpu` (driver error 209). Forcing the standard module-load path lets the intercept translate every kernel.

Additionally, `_patch_rope_avoid_gemm` rewrites RoPE's frequency computation `inv_freq @ position_ids` (a degenerate `K=1` outer-product GEMM) as a bit-identical broadcast multiply. That matmul otherwise dispatches to hipBLASLt, which has no gfx1250 solution for the fp32 outer-product shape. The rewrite is applied to both branches so equivalence is preserved.

### Checklist status (gfx950)

- [x] **Unmodified model runs on gfx950** — native/local branch passes; greedy output produced for all 3 prompts.
- [x] **Relevant kernels extracted** — 48 Inductor/Triton kernels captured under the run's `torchinductor_cache/` and `triton_cache/`.
- [x] **Transpilation succeeds on gfx950** — **38/38 kernels** transpile gfx1250→gfx950 cleanly (0 raise failures).
- [x] **Individual kernels run with synthetic data** — all transpiled kernels load and launch under the HotSwap stack during the whole-model pass.
- [x] **Transpiled kernels produce results within FP-noise (no NaNs)** — equivalence verdict **pass**; per-case logprob/KL/TVD diffs are exactly `0.0`.
- [x] **Whole model runs with transpiled kernels** — HotSwap branch passes, `proof_status: translated`, 38/38 success (proof cache: 12 hits / 26 misses).
- [~] **Throughput measured** — see caveat below; not a representative warm number.
- [ ] **All of the above on BOTH gfx942 AND gfx950** — gfx942 not tested here (hardware is MI350X/gfx950 only).

### Throughput caveat

The runner does a single pass, so the reported aggregate tok/s includes full `torch.compile` compilation time over only 16 tokens × 3 prompts: local `0.180 tok/s`, HotSwap `0.136 tok/s`. These are compile-dominated and **not** a meaningful steady-state throughput. A representative number needs a warmup pass + measured warm pass (as the gpt-oss runner does with `run1`/`run2`); this can be added if a hard token/s figure is needed for the checklist.

### Library-kernel override path (evaluated, not used)

The only pre-existing causal-LM config in the repo (`phi4_mini.json`) uses the eager + `gfx_override_target` path, which translates the rocBLAS/hipBLASLt GEMM kernels themselves. That path is not viable in this environment:

- The committed gfx1250 Tensile fixtures (`data/kernels/gfx1250_b0/…`) are a curated subset, not a full BLAS library. hipBLASLt's heuristic returned no gfx1250 solution for Llama's general GEMM shapes (aborted at `q_proj` with `HIPBLAS_STATUS_INVALID_VALUE`).
- rocBLAS derives its Tensile library filename from the **physical** device arch (gfx950) and ignores the `hipGetDeviceProperties` override used for library scoping, so it cannot be pointed at the gfx1250 lazy library (`Cannot read TensileLibrary.dat … for GPU arch : gfx950`).

The `torch.compile`/Triton path above is therefore the viable route on this stack.

### Result

```
## hotswap
- Passed: True
- Proof status: translated
- HotSwap proof: 38/38 success

## Equivalence
- Verdict: pass
- Cases: 3 total, 0 token-divergent  (per-step diffs 0.0)
```

### Pull request

Raised against `users/adespirl/non-embedding-model-groups` (authored `Ish Kool <ishkool2@amd.com>`):

- **Llama** — [AMD-ROCm-Internal/rocm-hotswap-testing#22](https://github.com/AMD-ROCm-Internal/rocm-hotswap-testing/pull/22): `runtime/pytorch/workloads/hf_causal_lm.py`, `data/pytorch/models/llama_3_2_3b_instruct.json`, `data/pytorch/models/llama_3_2_3b_instruct.prompts.json`.

**Comment by ishkool** [comment]
## gfx942 validation report — `meta-llama/Llama-3.2-3B-Instruct`

**TL;DR:** The model runs end-to-end through the PyTorch HotSwap runner on **gfx942**. The native/local branch passes and the **whole-model HotSwap run passes**: **38/38 Triton kernels transpiled** gfx1250→gfx942 (0 raise failures) and greedy generation is **numerically equivalent** to native (verdict `numerically_close`; 0 token-divergent across all 3 prompts, every sequence bit-identical in chosen tokens). This reuses the causal-LM enablement from the gfx950 bring-up (Llama commit `ef0af24c`: static KV cache under compile, autotuning + static-CUDA-launcher disabled, RoPE GEMM rewrite) with no new changes — only the transpiler target was retargeted to gfx942.

### Environment

| Component | Value |
|---|---|
| GPU | AMD Instinct **MI300X**, `gfx942` (wave64) |
| Container | `rocm/pytorch:latest`, ROCm **7.2.4** (`hip 7.2.53211`) |
| `llvm-project` (COMGR) | branch `hotswap` @ `f4f207cbfdddbd822c365bb9ef1eba0e3896c316` |
| `rocm-systems` (ROCR) | `users/mluecke/hotswap-compatibility` @ `26b07419c2e5985aa9874b4448190ccb6cc6286c` |
| `rocm-hotswap-testing` | `users/adespirl/non-embedding-model-groups` @ `cfc62d47` + Llama commit `ef0af24c` (`origin/users/ishkool2/llama-3.2-3b-hotswap`) |
| torch / triton | `2.12.0+rocm7.2` / `triton-rocm 3.6.0` |
| transformers | causal-LM (`AutoModelForCausalLM`), `sdpa` attention |
| Source target / target | `gfx1250:32` (wave32) → `gfx942` (wave64) |
| Precision / workload | `bfloat16`, 3 chat prompts, greedy decode, `max_new_tokens 16`, `top_logprobs 5` |

Llama-3.2-3B-Instruct is gated; weights were provided locally at `/home/AMD/ishkool2/llama/Llama-3.2-3B-Instruct` and passed via `--model-path`. Because the config declares `target_gfx: gfx950`, the run was launched with `--target-gfx gfx942` to retarget the transpiler onto this hardware.

### Runner enablement

Llama fits the existing generic `hf_causal_lm` workload — enabling it is a JSON config plus the causal-LM workload changes, not a new Python module. Config: `data/pytorch/models/llama_3_2_3b_instruct.json` (`workload: hf_causal_lm`, `source_target: gfx1250:32`, `target_gfx: gfx950`, `PYTORCH_COMPILE: 1`, equivalence `atol_logprob 5e-3` / `kl_tolerance 1e-3`) plus `llama_3_2_3b_instruct.prompts.json`.

A causal LM in eager mode dispatches entirely to ATen + rocBLAS/hipBLASLt library kernels and emits **no** JIT Triton kernels, so there is nothing for the transpiler to translate. To exercise the HotSwap path the model is run under `torch.compile` (TorchInductor → Triton), which routes fused pointwise/reduction ops (RMSNorm, RoPE apply, SiLU, softmax, residual adds) through Triton; those kernels are force-compiled for `gfx1250:32` and translated to `gfx942` on load. Matmuls (Q/K/V/O and MLP projections, `lm_head`) still dispatch to the native gfx942 BLAS backend — the same coverage model as the accepted DiT/FLUX runs.

The causal-LM workload changes carried in `hf_causal_lm.py` (from the gfx950 bring-up, commit `ef0af24c`) are what make that path work under HotSwap and apply identically on gfx942:

1. **Static KV cache under compile** (`cache_implementation="static"`) — HF `generate()` otherwise graph-breaks on its default dynamic cache and Inductor never compiles the decode graph.
2. **Disable all Inductor/Triton autotuning** — a single failing candidate in an autotune sweep would abort the whole kernel; disabling gives one deterministic config per kernel.
3. **Disable the static CUDA launcher** (`use_static_cuda_launcher=False`) — it bypasses `hipModuleLoadDataEx` (the API the HotSwap intercept hooks), which would load raw gfx1250 binaries on the target → `hipErrorNoBinaryForGpu` (Code 209). Forcing the standard module-load path lets the intercept translate every kernel.
4. **`_patch_rope_avoid_gemm`** — rewrites RoPE's `inv_freq @ position_ids` (a degenerate `K=1` outer-product GEMM with no gfx1250 BLAS solution) as a bit-identical broadcast multiply, applied to both branches so equivalence is preserved.

### Checklist status (gfx942)

- [x] **Unmodified model runs on gfx942** — native/local branch passes; greedy output produced for all 3 prompts.
- [x] **Relevant kernels extracted** — Inductor/Triton kernels captured under the run's `torchinductor_cache/` and `triton_cache/`.
- [x] **Transpilation succeeds on gfx942** — **38/38 kernels** transpile gfx1250→gfx942 cleanly (0 raise failures; proof cache 12 hits / 26 misses).
- [x] **Individual kernels run with synthetic data** — all transpiled kernels load and launch under the HotSwap stack during the whole-model pass.
- [x] **Transpiled kernels produce results within FP-noise (no NaNs)** — equivalence verdict **pass / numerically_close**; 0 NaN/Inf in logprob streams.
- [x] **Whole model runs with transpiled kernels** — HotSwap branch passes, `proof_status: translated`, 38/38 success.
- [~] **Throughput measured** — single-pass, compile-dominated (see caveat); not a representative warm number.
- [x] **BOTH gfx942 AND gfx950** — gfx950 previously validated (38/38, equivalent); **gfx942 now validated here (38/38, numerically_close)**.

### Numerical equivalence (local vs HotSwap)

- **Verdict:** PASS (`numerically_close`); 3/3 cases passed, **0 token-divergent**, every case `sequence_equal = true`.
- **Max chosen-token logprob diff:** `1.48e-3` (tolerance `atol_logprob = 5e-3`).
- **Max forward KL:** `1.77e-4` (tolerance `kl_tolerance = 1e-3`); max TVD `1.47e-3`.
- **NaN logprobs:** 0 on all cases.

| Case | Status | chosen-logprob max_abs_diff | KL max | TVD max |
|------|--------|----------------------------:|-------:|--------:|
| capital_france | equivalent | 0.0 | 0.0 | 0.0 |
| identity_hotswap | numerically_close | 2.02e-4 | 4.46e-6 | 1.44e-4 |
| add_two_numbers | numerically_close | 1.48e-3 | 1.77e-4 | 1.47e-3 |

### Throughput caveat

The runner does a single pass, so aggregate tok/s includes full `torch.compile` compilation over only 16 tokens × 3 prompts: local `0.173 tok/s`, HotSwap `0.170 tok/s`. These are compile-dominated and **not** a meaningful steady-state throughput (same caveat as the gfx950 run). A representative number needs a warmup + measured warm pass.

### Result

```
## hotswap
- Passed: True
- Proof status: translated
- HotSwap proof: 38/38 success

## Equivalence
- Verdict: pass  (numerically_close)
- Cases: 3 total, 0 token-divergent
```

The gfx942 leg of #75 is now complete alongside the existing gfx950 pass — both hardware targets pass end-to-end (38/38 kernels, numerically equivalent greedy generation).
